File: Python/BinarySearchTree.py
from StatTracker import StatTracker
import random

class Node(object):
    def __init__(self, value):
        self.left = None
        self.right = None
        self.value = value

    def __str__(self):
        return str(self.value)


class BinarySearchTree(object):
    def __init__(self, root_value):
        # Initialize a tree with the root node value of (root_value)
        self.root = self.insert(None, root_value)

    def _create_node(self, value):
        return Node(value)

    def insert(self, root, value):
        # If root is None, we have found the location where we wish to insert
        # our node
        if root == None:
            return self._create_node(value)
        else:
            # Traverse the LEFT side of the tree
            if value <= root.value:
                root.left = self.insert(root.left, value)
            # Traverse the RIGHT side of the tree
            else:
                root.right = self.insert(root.right, value)
            return root

    def search(self, value, node, stats=None):
        # Search the tree for a given value.
        # Track stats for the search, if requested
        if stats != None:
            stats.tick()
        # If the current node doesn't exist, the value is not in our tree
        if node == None:
            print "No value '%s' found in tree" % value
            return
        # If the current node's value is the value we are looking for, return it
        elif node.value == value:
            if stats != None:
                stats.done()
                print ( " --------------------------- \n" +
                        "Node found!\n" +
                        " -- %s nodes visited\n" % stats.nodes_visited +
                        " -- %s time elapsed\n" % stats.total_time +
                        " -- node value: %s\n" % node.value +
                        " -- node left child: %s\n" % node.left +
                        " -- node right child: %s\n" % node.right +
                        " --------------------------- ")

            return node
        elif value < node.value:
            return self.search(value, node.left, stats)
        else:
            return self.search(value, node.right, stats)


if __name__ == '__main__':
    max = 1000000
    tree = BinarySearchTree(random.randint(0, max))
    for i in range(max/2):
        tree.insert(tree.root, random.randint(0, max))

    st = StatTracker()

    for i in range(10):
        tree.search(random.randint(0, max), tree.root, st)
        st.reset()

File: Java/BinarySearchTree.java

class Node
{
    private int value;
    private boolean isRoot;
    private Node leftChild;
    private Node rightChild;

    public Node(int val)
    {
        value = val;
        isRoot = false;
    }

    public void setAsRoot()
    {
        isRoot = true;
    }

    public boolean isRoot()
    {
        return isRoot;
    }

    public int getValue()
    {
        return value;
    }

    public void setValue(int val)
    {
        value = val;
    }

    public Node getLeftChild()
    {
        return leftChild;
    }

    public void setLeftChild(Node child)
    {
        leftChild = child;
    }

    public Node getRightChild()
    {
        return rightChild;
    }

    public void setRightChild(Node child)
    {
        rightChild = child;
    }
}


class BST
{
    private Node root;

    public BST(int rootVal)
    {
        // Insert a new Node into the tree with (null) parent and value of
        // (rootVal)
        Node empty = new Node(0);
        empty.setAsRoot();
        root = insert(empty, rootVal);
    }

    private Node createNode(int value)
    {
        Node node = new Node(value);
        return node;
    }

    public Node getRoot()
    {
        return root;
    }

    public Node insert(Node root, int val)
    {
        // If root is null, we have found the location where we wish to insert
        // our node
        if (root == null)
        {
            return createNode(val);
        }
        else
        {
            // Traverse the left subtree
            if (val <= root.getValue())
            {
                root.setLeftChild(insert(root.getLeftChild(), val));
            }
            // Traverse the right subtree
            else
            {
                root.setRightChild(insert(root.getRightChild(), val));
            }

            return root;
        }
    }

    public void search(int value, Node node)
    {
        // IMPLEMENT ME!
    }
}


public class BinarySearchTree
{
    public static void main(String[] args)
    {
        // Create a BST and play with it
        BST tree = new BST(5);
        Node root = tree.getRoot();
        tree.insert(root, 0);
        tree.insert(root, 1);
        tree.insert(root, 2);
        tree.insert(root, 3);
        tree.insert(root, 4);
        tree.insert(root, 6);
        tree.insert(root, 7);
        tree.insert(root, 8);
        tree.insert(root, 9);
    }
}
